smart_open/asb.py

```diff
@@ -0,0 +1,417 @@
+        if size == 0:
+            return b''
+        elif size < 0:
+            self._current_pos = self._size
+            return self._read_from_buffer() + self._raw_reader.read()
+
+        #
+        # Return unused data first
+        #
+        if len(self._current_part) >= size:
+            return self._read_from_buffer(size)
+
+        #
+        # If the stream is finished, return what we have.
+        #
+        if self._eof:
+            return self._read_from_buffer()
+
+        #
+        # Fill our buffer to the required size.
+        #
+        self._fill_buffer(size)
+        return self._read_from_buffer(size)
+
+    def read1(self, size=-1):
+        """This is the same as read()."""
+        return self.read(size=size)
+
+    def readinto(self, b):
+        """Read up to len(b) bytes into b, and return the number of bytes
+        read."""
+        data = self.read(len(b))
+        if not data:
+            return 0
+        b[:len(data)] = data
+        return len(data)
+
+    def readline(self, limit=-1):
+        """Read up to and including the next newline.  Returns the bytes read."""
+        if limit != -1:
+            raise NotImplementedError('limits other than -1 not implemented yet')
+        the_line = io.BytesIO()
+        while not (self._eof and len(self._current_part) == 0):
+            #
+            # In the worst case, we're reading the unread part of self._current_part
+            # twice here, once in the if condition and once when calling index.
+            #
+            # This is sub-optimal, but better than the alternative: wrapping
+            # .index in a try..except, because that is slower.
+            #
+            remaining_buffer = self._current_part.peek()
+            if self._line_terminator in remaining_buffer:
+                next_newline = remaining_buffer.index(self._line_terminator)
+                the_line.write(self._read_from_buffer(next_newline + 1))
+                break
+            else:
+                the_line.write(self._read_from_buffer())
+                self._fill_buffer()
+        return the_line.getvalue()
+
+    #
+    # Internal methods.
+    #
+    def _read_from_buffer(self, size=-1):
+        """Remove at most size bytes from our buffer and return them."""
+        # logger.debug('reading %r bytes from %r byte-long buffer', size, len(self._current_part))
+        size = size if size >= 0 else len(self._current_part)
+        part = self._current_part.read(size)
+        self._current_pos += len(part)
+        # logger.debug('part: %r', part)
+        return part
+
+    def _fill_buffer(self, size=-1):
+        size = size if size >= 0 else self._current_part._chunk_size
+        while len(self._current_part) < size and not self._eof:
+            bytes_read = self._current_part.fill(self._raw_reader)
+            if bytes_read == 0:
+                logger.debug('reached EOF while filling buffer')
+                self._eof = True
```

Comment:
Make this method return something instead of passing this information via a member variable.